Optimize shared texture creations

ze2j opened the pull request:
Fix [#98733](https://github.com/godotengine/godot/issues/98733)

This PR removes an unnecessary copy of `Texture::slice_trackers`, as it is always cleared afterward.

Build: scons platform=linuxbsd optimize=speed_trace debug_symbols=yes
Profiler: Tracy 
Test project: see MRP in  [#98733](https://github.com/godotengine/godot/issues/98733)

Results with master branch:
![master](https://github.com/user-attachments/assets/64fe6870-cb04-4b63-8c08-ab397a5ede47)

Results with this PR:
![pr](https://github.com/user-attachments/assets/379feeb0-0e9a-4075-bbc7-4f63403d781f)

clayjohn commented on servers/rendering/rendering_device.cpp:
```diff
@@ -903,7 +903,7 @@ RID RenderingDevice::texture_create_shared(const TextureView &p_view, RID p_with
 	// Create view.
 
-	Texture texture = *src_texture;
+	Texture texture = src_texture->duplicate_as_shared_texture();
 	texture.shared_fallback = nullptr;
```
I guess `duplicate_as_shared_resource` should also avoid copying `shared_fallback` so we don't have to manually set it to nullptr right after doing the copy

clayjohn commented on servers/rendering/rendering_device.cpp:
```diff
@@ -1055,7 +1053,7 @@ RID RenderingDevice::texture_create_shared_from_slice(const TextureView &p_view,
 	}
 
-	Texture texture = *src_texture;
+	Texture texture = src_texture->duplicate_as_shared_texture();
```
I find it a bit surprising that we don't call `texture.slice_trackers.clear();` here. 

CC @DarioSamo who added the call in `texture_create_shared()`. Do you think this was missed by mistake? Or is there a reason to maintain the `slice_tracker` when creating a shared texture from a slice?

ze2j replied:
The clear is made in `_texture_make_mutable`

DarioSamo commented:
I think `slice_trackers` should be turned into a pointer instead and allocate it on demand only if a shared texture is created. The HashMap currently occupies 48 bytes per texture and largely goes unused in the vast majority of cases. Then when the copy is made, the simple solution is to just set it to a nullptr. That way there'll be no cost from copying the contents as is and we get a reduced memory footprint overall, for the small extra cost of an indirection.

I'm not a big fan of the proposed solution as it adds another function to maintain and keep track of when the copy is made.

ze2j commented:
I agree and I am not a big fan of my solution too : ) I am going to rework my PR with your solution, it's a nice idea.

ze2j committed:
Optimize shared texture creation

Texture::slice_trackers is now a pointer and is allocated on demand only when a shared texture is created.
This makes copying Texture significantly faster.

ze2j commented:
Here is the new implementation following @DarioSamo suggestion. Sorry for the delay.

DarioSamo commented:
I'm probably in need of this so I will test it again and confirm if it works.

DarioSamo reviewed (approved):
Tested and confirmed to be working. Gives me a very nice time reduction on the creation of reflection probe atlases.

Calinou commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5844,9 +5844,12 @@ bool RenderingDevice::_texture_make_mutable(Texture *p_texture, RID p_texture_id
 					// Slice texture.
-					HashMap<Rect2i, RDG::ResourceTracker *>::ConstIterator draw_tracker_iterator = owner_texture->slice_trackers.find(p_texture->slice_rect);
+					if (owner_texture->slice_trackers == nullptr) {
+						owner_texture->slice_trackers = memnew((HashMap<Rect2i, RDG::ResourceTracker *>));
```
Do we need the double parentheses on each side here?

ze2j replied:
Yes it's definitely mandatory, it doesn't build without it:
`error: macro "memnew passed 2 arguments,  but takes just 1"`

MarianoGnu commented:
> Gives me a very nice time

Out of curiosity, the profiling shown at the top of the discussion was from the original implementation or of this rework, has the result improved in speed beside memory footprint?

ze2j commented:
@MarianoGnu  the profiling shown at the top of the discussion was from the original implementation. Here is the new one: 
![pr_new_impl](https://github.com/user-attachments/assets/d5c531c4-8780-493b-acc9-c4ab42eef87b)

Repiteo commented:
Thanks!
